File: eosknight.py
#!/usr/bin/env python
# coding: utf-8

from __future__ import division
import urllib
import requests
import json
import time
import sys

# 计算击杀数
def calcKillCount(defense, hp, attack, time = 864000000000):
    damage_per_min = 25 - (25 * defense) / (defense + 1000)
    alive_sec = (60 * hp) / damage_per_min
    if time < alive_sec:
        alive_sec = time
    current_kill_count = (attack * alive_sec) / (60 * 200)
    return current_kill_count


# 计算存活时间
def calcKillAliveTime(defense, hp):
    damage_per_min = 25 - (25 * defense) / (defense + 1000)
    alive_sec = (60 * hp) / damage_per_min
    return alive_sec


# 计算多个骑士最大存活时间
def getKnightsMaxTime(encodedName):
    url = 'https://api.eosnewyork.io/v1/chain/get_table_rows'

    data_json = {'json': True, 'code': 'eosknightsio', 'scope': 'eosknightsio', 'table': 'knight', 'table_key': '',
                 'key_type': 'i64', 'lower_bound': encodedName, 'index_position': 1, 'limit': 1}

    headers = {'content-type': 'application/json; charset=UTF-8'}
    data = json.dumps(data_json, ensure_ascii=False).encode('utf8')
    try:
        r = requests.post(url, data=data, headers=headers)
        ans = r.json()['rows'][0]
        t1 = calcKillAliveTime(ans['rows'][0]['defense'], ans['rows'][0]['hp'])
        t2 = calcKillAliveTime(ans['rows'][1]['defense'], ans['rows'][1]['hp'])
        t3 = calcKillAliveTime(ans['rows'][2]['defense'], ans['rows'][2]['hp'])
        return max(t1, t2, t3)

    except Exception, e:
        print Exception, ":", e
        return 0


# 获取骑士当前时间
def getKnightsCurrTime(encodedName):
    url = 'https://api.eosnewyork.io/v1/chain/get_table_rows'

    data_json = {'json': True, 'code': 'eosknightsio', 'scope': 'eosknightsio', 'table': 'player', 'table_key': '',
                 'key_type': 'i64', 'lower_bound': encodedName, 'index_position': 1, 'limit': 1}

    headers = {'content-type': 'application/json; charset=UTF-8'}
    data = json.dumps(data_json, ensure_ascii=False).encode('utf8')
    try:
        r = requests.post(url, data=data, headers=headers)
        ans = r.json()['rows'][0]['last_rebirth']
        return int(time.time()) - 1500000000 - ans
    except Exception, e:
        print Exception, ":", e
        return 0


# 获取宠物远征是否可以回来
def getPetExpedition(account, encodedName):
    url = 'https://api.eosnewyork.io/v1/chain/get_table_rows'

    data_json = {'json': True, 'code': 'eosknightsio', 'scope': 'eosknightsio', 'table': 'petexp', 'table_key': '',
                 'key_type': 'i64', 'lower_bound': encodedName, 'index_position': 1, 'limit': 1}

    headers = {'content-type': 'application/json; charset=UTF-8'}
    data = json.dumps(data_json, ensure_ascii=False).encode('utf8')
    try:
        r = requests.post(url, data=data, headers=headers)
        pets = r.json()['rows'][0]['rows']
        # print pets
        curr = int(time.time()) - 1500000000
        for pet in pets:
            if pet['isback'] == 0:  # 处于远征状态
                # print curr,pet['end']
                if curr >= pet['end'] and curr - pet['end'] < 600:  # 可以回来了
                    print '%s 宠物远征结束了' % account
                    return '%s 宠物远征结束了' % account

        return ''
    except Exception, e:
        print Exception, ":", e
        return ''


# 微信推送消息
def sendToWechat(sckey, msg):
    url = 'https://sc.ftqq.com/%s.send?text=%s' % (
        sckey, urllib.quote(msg))
    r = requests.get(url)


# 检查是否该复活了
def checkKnight(account, encodedName):
    maxTime = getKnightsMaxTime(encodedName)
    currTime = getKnightsCurrTime(encodedName)
    print "maxtime,currTime", maxTime, currTime
    if currTime >= maxTime and currTime - maxTime < 600:
        global checkTime
        checkTime = 30
        return '%s 该复活啦~\n' % (account)
    else:
        global checkTime
        checkTime = 5
        return ''


# 获取字母的序号
def getLetterNumber(letter):
    strMap = ".12345abcdefghijklmnopqrstuvwxyz"
    # .=>0*16
    # 1=>1*16
    # 2=>2*32
    # a=>6*16
    # z=>31*16
    return strMap.index(letter) * 16


# 获取账号的big number
def getEncodedAccount(account):
    try:
        account = account.rjust(12, '.')
        ret = 0
        index = 0
        # ascii 49=>1  53=>5  97=>a  122=>z
        for letter in reversed(account):
            int = getLetterNumber(letter)
            ret = ret + pow(32, index) * int
            index = index + 1
        return ret
    except:
        print "账号出错！请检查账号！"
        sys.exit(1)


if __name__ == '__main__':
    global checkTime
    checkTime = 5
    ################################################
    sckey = ''  # 你的key
    accounts = {
        'linzhanjie51'
    }
    ################################################
    encodedAccounts = {}
    for account in accounts:
        encodedAccounts[account] = getEncodedAccount(account)
    # print encodedAccounts

    while True:
        try:
            str = ''
            for acc in accounts:
                encAcc = encodedAccounts[acc]
                # print acc,encAcc
                str += checkKnight(acc, encAcc)
                str += getPetExpedition(acc, encAcc)
            print str
            if len(str) > 12:
                sendToWechat(sckey, str)
            print '========' + time.strftime('%Y.%m.%d %H:%M:%S', time.localtime(time.time())) + '============'
            time.sleep(60 * checkTime)  # 每5分钟循环一次
        except Exception as e:
            print "有错误！！", e
            time.sleep(20)  # 如果发生错误，延迟20秒再次执行

File: eosknight.js
const axios = require('axios');

// 账号参数
const sckey = '';
const account = 'linzhanjie51';

// 拉表格
const getTableData = async table => {
    return new Promise((resolve, reject) => {
        const encodedName = getEncodeAccount(account);
        const url = 'https://api.eosnewyork.io/v1/chain/get_table_rows';
        const data = {
            json: true,
            code: 'eosknightsio',
            scope: 'eosknightsio',
            table: table,
            table_key: '',
            key_type: 'i64',
            lower_bound: encodedName,
            index_position: 1,
            limit: 1
        };
        const headers = {
            'content-type': 'application/json; charset=UTF-8'
        }
        axios({
            method: 'POST',
            url: url,
            headers: headers,
            data: data,
            responseType: 'json'
        }).then((r) => {
            const responseData = r.data.rows[0];
            resolve(responseData);
        }).catch((error) => {
            if (error.response) {
                console.log(error.response.data);
                reject(error);
            }
        });
    });
}

// 算杀人数，目前没用到
function calcKillCount(defense, hp, attack, time = 864000000000) {
    let damage_per_min = 25 - (25 * defense) / (defense + 1000);
    let alive_sec = (60 * hp) / damage_per_min;
    if (time < alive_sec) {
        alive_sec = time;
    }
    let current_kill_count = (attack * alive_sec) / (60 * 200);
    return current_kill_count;
}

function calcKillAliveTime(defense, hp) {
    let damage_per_min = 25 - (25 * defense) / (defense + 1000);
    let alive_sec = (60 * hp) / damage_per_min;
    return alive_sec;
}

// 计算最大时间
async function getKnightsMaxTime() {
    const dataObj = await getTableData('knight');
    const knights = ['knight', 'archer', 'mage'];
    const knightsObj = {};
    knights.map((role, index) => {
        index +=1 ;
        const obj = dataObj['rows'].find(function(x) {
            return x.type === index;
        });
        knightsObj[`${role}Obj`] = obj;
    });
    const knightMaxTime = calcKillAliveTime(knightsObj.knightObj.defense, knightsObj.knightObj.hp);
    const archerMaxTime = calcKillAliveTime(knightsObj.archerObj.defense, knightsObj.archerObj.hp);
    const mageMaxTime = calcKillAliveTime(knightsObj.mageObj.defense, knightsObj.mageObj.hp);
    return Math.max(knightMaxTime, archerMaxTime, mageMaxTime);
    // console.log(Math.max(knightMaxTime, archerMaxTime, mageMaxTime));
}

// 获取当前时间
async function getKnightsCurrTime() {
    const dataObj = await getTableData('player');
    const lastRebirthTime = dataObj.last_rebirth;
    const currTime = Date.now() / 1000 - 1500000000 - lastRebirthTime;
    return currTime;
}

// 判断宠物远征
async function checkPetExpedition() {
    const dataObj = await getTableData('petexp');
    const pets = dataObj.rows;
    const currTime = Date.now() / 1000 - 1500000000;
    pets.map((pet, index) => {
        if (pet.isback === 0 && currTime >= pet.end && currTime - pet.end < 600) {
            msg = '宠物远征结束啦';
        }
    });
    if (msg === '') {
        console.log('好像没有远征结束的');
    }
}

// 判断能否复活
async function checkRebirth() {
    let maxTime, currTime;
    await getKnightsMaxTime().then(res => {maxTime = res});
    await getKnightsCurrTime().then(res => {currTime = res});
    if (currTime >= maxTime && currTime - maxTime < 600) {
        msg = '可以复活啦';
    } else {
        msg = '';
        console.log('好像没有可以复活的');
    }
}

// 获取字母序号
function getLetterNumber(letter) {
    const strMap = '.12345abcdefghijklmnopqrstuvwxyz';
    return strMap.indexOf(letter) * 16
}

// 获取账号的big number（天晓得这是啥）
function getEncodeAccount(account) {
    try {
        account = account.padStart(12, '.');
        let ret = 0;
        let reverseAccountArray = account.split('').reverse();
        reverseAccountArray.map((currentLetter, index) => {
            let letterNumber = getLetterNumber(currentLetter);
            ret = ret + Math.pow(32, index) * letterNumber;
        });
        return ret;
    }
    catch (e) {
        console.log(e);
        console.log('账号出错啦！');
    }
}

// 微信推送
function sendToWechat(msg) {
    let url = `https://sc.ftqq.com/${sckey}.send?text=${msg}`;
    axios({
        method: 'POST',
        url: url,
    }).then(() => {msg = ''});
}

// sleep
const sleep = (m) => {
    return new Promise((resolve, reject) => {
        setTimeout(resolve, 1000 * 60 * m);
    }).catch(e => reject(e));
}

async function checkAction() {
    try {
        let getHour = new Date().getUTCHours() + 8;
        if (getHour >= 1 && getHour <= 9 ) {
            checkTimeStep = 60;
        } else {
            await checkRebirth();
            if (msg !== '') {
                console.log(`消息：${msg}`);
                await sendToWechat(encodeURI(msg));
            }
            await sleep(1);
            await checkPetExpedition();
            if (msg !== '') {
                console.log(`消息：${msg}`);
                await sendToWechat(encodeURI(msg));
            }
        }
    }
    catch (e) {
        console.log(e);
        msg = '出错了！快看日志！';
        await sendToWechat(encodeURI(msg));
    }
}

let msg = '';
checkAction()
setInterval(() => {checkAction()}, 1000 * 60 * 5);
// checkAction();
// sendToWechat(encodeURI('中文试试'));
